### PR TITLE
implemented NamedCaffeineCache in terms of AsyncCache

### PR DESCRIPTION
**Pull request by gurkankaymak**
implemented NamedCaffeineCache in terms of AsyncCache instead of Cache interface as discussed #9615

### CODE REVIEW

**Comment by ben-manes** [comment]
There are default methods you may want to implement. 
```java
@NonNull
default CompletableFuture<Map<K, V>> getAll(@NonNull Iterable<? extends @NonNull K> keys,
    @NonNull Function<Iterable<? extends @NonNull K>, @NonNull Map<K, V>> mappingFunction) {
  // This method was added & implemented in version 2.8.0
  throw new UnsupportedOperationException();
}

@NonNull
default CompletableFuture<Map<K, V>> getAll(@NonNull Iterable<? extends @NonNull K> keys,
    @NonNull BiFunction<Iterable<? extends @NonNull K>, Executor, CompletableFuture<Map<K, V>>> mappingFunction) {
  // This method was added & implemented in version 2.8.0
  throw new UnsupportedOperationException();
}
```

**Comment by gurkankaymak** [comment]
These methods do not exist for the version that play framework currently uses

Screen Shot 2019-09-10 at 20 39 10

**Comment by ben-manes** [comment]
Can you upgrade the version and delegate? I’m surprised to see a custom implementation of the logic was needed.

File: cache/play-caffeine-cache/src/main/java/play/cache/caffeine/NamedCaffeineCache.java [hunk 1]
```diff
@@ -4,23 +4,32 @@
 
 package play.cache.caffeine;
 
+import com.github.benmanes.caffeine.cache.AsyncCache;
 import com.github.benmanes.caffeine.cache.Cache;
-import com.github.benmanes.caffeine.cache.Policy;
-import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import org.checkerframework.checker.nullness.qual.NonNull;
```
Comment (PromanSEW) [inline cache/play-caffeine-cache/src/main/java/play/cache/caffeine/NamedCaffeineCache.java]: You can use `javax.annotation.Nonnull` which is already imported

**Comment by gurkankaymak** [comment]
> Thanks, @gurkankaymak.
> 
> One missing bit here is to have `CaffeineCacheApi` implemented not in terms of the synchronous cache, but instead use `NamedCaffeineCache` directly where possible. Not sure if your idea is to address this on #9615, though.

Hi @marcospereira yes I was planning to address this on #9615, currently only synch functionality is used, my plan is to make this pr simple implement the `NamedCaffeineCache` in terms of `AsyncCache` while retaining the same functionality, and update the #9615 on top of this pr and make use of the async functionality

**Comment by gurkankaymak** [comment]
> Can you upgrade the version and delegate? I’m surprised to see a custom implementation of the logic was needed.

isn't the latest version of `AsyncCache` throws UnsupportedOperationException?
play framework uses version 2.8.0

**Comment by ben-manes** [comment]
The interface methods were added in a semver compatible way, but the Caffeine  implementation overrides them with the desired logic. This was to avoid a major version bump and not break a outside implementors like Play’s which could add the delegation at its own pace.

**Comment by gurkankaymak** [comment]
> The interface methods were added in a semver compatible way, but the Caffeine implementation overrides them with the desired logic. This was to avoid a major version bump and not break a outside implementors like Play’s which could add the delegation at its own pace.

ok got it, then I'm updating to delegate, thanks for the explanation

File: cache/play-caffeine-cache/src/main/java/play/cache/caffeine/NamedCaffeineCache.java (hunk 1 above)
Comment (gurkankaymak) [reply cache/play-caffeine-cache/src/main/java/play/cache/caffeine/NamedCaffeineCache.java]: you're right this was auto-imported and I did not notice, thanks for your input

File: cache/play-caffeine-cache/src/test/java/play/cache/caffeine/NamedCaffeineCacheSpec.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    String value1 = "value1";
+    String key2 = "key2";
+    String value2 = "value2";
+    cache.put(key1, CompletableFuture.completedFuture(value1));
+    cache.put(key2, CompletableFuture.completedFuture(value2));
+    Set<String> keys = new HashSet<>(Arrays.asList(key1, key2));
+
+    CompletableFuture<Map<String, String>> futureResult =
+        cache.getAll(
+            keys, (missingKeys, executor) -> CompletableFuture.completedFuture(new HashMap<>()));
```
Comment (marcospereira) [inline cache/play-caffeine-cache/src/test/java/play/cache/caffeine/NamedCaffeineCacheSpec.java]: Use `Collections.emptyMap` instead of `new HashMap`.

Comment (marcospereira) [reply cache/play-caffeine-cache/src/test/java/play/cache/caffeine/NamedCaffeineCacheSpec.java]: This is a nitpick since the code is in tests.

**Pull request merged**